transport/whisper/types/whisper.go

```diff
@@ -0,0 +1,61 @@
+package whispertypes
+
+import (
+	"crypto/ecdsa"
+	"time"
+)
+
+// Whisper represents a dark communication interface through the Ethereum
+// network, using its very own P2P communication layer.
+type Whisper interface {
+	PublicWhisperAPI() PublicWhisperAPI
+	NewMessageStore() MessageStore
+
+	// MinPow returns the PoW value required by this node.
+	MinPow() float64
+	// BloomFilter returns the aggregated bloom filter for all the topics of interest.
+	// The nodes are required to send only messages that match the advertised bloom filter.
+	// If a message does not match the bloom, it will tantamount to spam, and the peer will
+	// be disconnected.
+	BloomFilter() []byte
+	// SetTimeSource assigns a particular source of time to a whisper object.
+	SetTimeSource(timesource func() time.Time)
+	// GetCurrentTime returns current time.
+	GetCurrentTime() time.Time
+
+	// SelectedKeyPairID returns the id of currently selected key pair.
+	// It helps distinguish between different users w/o exposing the user identity itself.
+	SelectedKeyPairID() string
+	// GetPrivateKey retrieves the private key of the specified identity.
+	GetPrivateKey(id string) (*ecdsa.PrivateKey, error)
+
+	SubscribeEnvelopeEvents(events chan<- EnvelopeEvent) Subscription
+
+	// AddKeyPair imports a asymmetric private key and returns a deterministic identifier.
+	AddKeyPair(key *ecdsa.PrivateKey) (string, error)
+	// DeleteKeyPair deletes the specified key if it exists.
+	DeleteKeyPair(key string) bool
+	// SelectKeyPair adds cryptographic identity, and makes sure
+	// that it is the only private key known to the node.
+	SelectKeyPair(key *ecdsa.PrivateKey) error
+	AddSymKeyDirect(key []byte) (string, error)
+	AddSymKeyFromPassword(password string) (string, error)
+	DeleteSymKey(id string) bool
+	GetSymKey(id string) ([]byte, error)
+
+	Subscribe(f Filter) (string, error)
+	GetFilter(id string) Filter
```

Comment:
The Nimbus implementation already has a `getFilterMessages` implemented. You retrieve messages through the filterId and not by passing the actual filter. The API only offers this call to retrieve messages, or via a callback. This way there is no need for a `GetFilter()` call.

Reply:
That's a good point. `status-protocol-go` seems to be using `GetFilter` (e.g. [here](https://github.com/status-im/status-protocol-go/blob/feature%2Fstatus-nim/transport/whisper/whisper_service.go#L196)), but I can probably refactor that to go through `PublicWhisperAPI.GetFilterMessages`.

Reply:
I replaced all but one `GetFilter` method calls, because the status-protocol-go method is using it to return a `map[Filter][]whispertypes.ReceivedMessage`. That method is marked as deprecated (probably being used still by status-react), so we will be able to drop it.

Reply:
@kdeme One hitch here is that `Filter.Retrieve` returns `whisper.RetrievedMessages` while `GetFilterMessages` returns a `whisper.Message` (which is the RPC representation).

Reply:
So considering you now replace all `Retrieve` with `GetFilterMessages`. Can't `RetrievedMessage` go away completely? Or do we still need the `ToWhisperMessage` call?

Reply:
Yeahh, I thought the same (and actually removed `ToWhisperMessage`), but we still need the type [here](https://github.com/status-im/status-protocol-go/blob/60301521af104e95579e38a61e361eb84350f454/transport/whisper/gethbridge/message_store_wrapper.go#L29)